Fixes for the PETSc 3.17 series

cticenhour opened the pull request:
## Reason
Updating to newer versions of PETSc enables easier maintenance and new features. But MOOSE requires some fixes to enable updating PETSc. 

## Design
Fixup MOOSE build and operation deficiencies for PETSc 3.17 series.

## Impact
Easier upgrade path for PETSc.

moosebuild commented on petsc:
```diff
@@ -1 +1 @@
-Subproject commit 477e44bbb558b1357d86363677accbb4bcdfaabc
+Subproject commit 9785d7c2f01a8ac390025cac12c784f374b5b0a5
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1237604/) on d7519fc wanted to post the following:

A change of the following file(s) triggered this check:

  conda/petsc

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1237604/) on d7519fc wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1237695/) on 3d71950 wanted to post the following:

A change of the following file(s) triggered this check:

  conda/petsc

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1237695/) on 3d71950 wanted to post the following:

A change of the following file(s) triggered this check:

  scripts/configure_petsc.sh

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml

The PETSc submodule or configuration was changed but the conda build config was not

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1237695/) on 3d71950 wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1237717/) on 3d71950 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/22723/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1238954/) on 2c69b0a wanted to post the following:

A change of the following file(s) triggered this check:

  conda/petsc

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1238954/) on 2c69b0a wanted to post the following:

A change of the following file(s) triggered this check:

  scripts/configure_petsc.sh

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml

The PETSc submodule or configuration was changed but the conda build config was not

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1238954/) on 2c69b0a wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1238954/) on 2c69b0a wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1239118/) on 2018727 wanted to post the following:

A change of the following file(s) triggered this check:

  conda/petsc

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1239118/) on 2018727 wanted to post the following:

A change of the following file(s) triggered this check:

  petsc

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml

The PETSc submodule or configuration was changed but the conda build config was not

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1239118/) on 2018727 wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1239118/) on 2018727 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

cticenhour committed:
Add support for new PtAP algorithm petsc option as of 3.17

Refs #22723

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1239650/) on 1c505bd wanted to post the following:

A change of the following file(s) triggered this check:

  conda/petsc

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1239650/) on 1c505bd wanted to post the following:

A change of the following file(s) triggered this check:

  petsc

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml

The PETSc submodule or configuration was changed but the conda build config was not

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1239650/) on 1c505bd wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1239650/) on 1c505bd wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1239650/) on 1c505bd wanted to post the following:

A change of the following file(s) triggered this check:

  conda/petsc

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1239650/) on 1c505bd wanted to post the following:

A change of the following file(s) triggered this check:

  petsc

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml

The PETSc submodule or configuration was changed but the conda build config was not

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1239650/) on 1c505bd wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1239650/) on 1c505bd wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

cticenhour committed:
Fixup unused value warning from PetscOptionsEnd

This was introduced for GCC compilers with PETSc 3.17 and has been fixed in 3.18, so the ignore pragmas can be removed when we update to that release.

Refs #22723

cticenhour committed:
Fixup deprecated Tao methods as of PETSc 3.17 in stochastic tools and optimization modules

Also fix some typos

Refs #22723

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1239756/) on 78012d7 wanted to post the following:

A change of the following file(s) triggered this check:

  conda/petsc

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1239756/) on 78012d7 wanted to post the following:

A change of the following file(s) triggered this check:

  petsc

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml

The PETSc submodule or configuration was changed but the conda build config was not

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1239756/) on 78012d7 wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1239756/) on 78012d7 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

loganharbour commented:
Throwing the container recipe on here just to see if it works.

lindsayad commented:
We want to do 3.17 as opposed to 3.18? I don't know exactly when the automated `-options_left` parameter got applied in PETSc but it's super nice

cticenhour commented:
The choice of 3.17 was mostly precedent based. We've lagged one minor revision behind for a while. If there's enough interest in jumping up to 3.18, it's pretty easy to reconfigure this PR.

lindsayad commented:
Eh, 3.17 is fine

cticenhour commented:
@oanaoana do you have an opinion here?

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1270138/) on 264e8c7 wanted to post the following:

A change of the following file(s) triggered this check:

  conda/petsc

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1270138/) on 264e8c7 wanted to post the following:

A change of the following file(s) triggered this check:

  petsc

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml

The PETSc submodule or configuration was changed but the conda build config was not

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1270138/) on 264e8c7 wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1270138/) on 264e8c7 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1270573/) on 79541fb wanted to post the following:

A change of the following file(s) triggered this check:

  petsc

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml

The PETSc submodule or configuration was changed but the conda build config was not

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1270573/) on 79541fb wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1270573/) on 79541fb wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1284947/) on d8ecbcd wanted to post the following:

A change of the following file(s) triggered this check:

  conda/petsc

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1284947/) on d8ecbcd wanted to post the following:

A change of the following file(s) triggered this check:

  petsc

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml

The PETSc submodule or configuration was changed but the conda build config was not

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1284947/) on d8ecbcd wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1284947/) on d8ecbcd wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1350603/) on 12d4920 wanted to post the following:

A change of the following file(s) triggered this check:

  conda/petsc

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

cticenhour commented:
Changing the scope of this PR to not include the submodule update for 3.17. It's been long enough since I tried to take this on that jumping to 3.18 seems a better option. I will merge the code fixes for the 3.17 and later changes and move on to a new PR with 3.18.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1350777/) on 04a57eb wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>b3bfe3</th>
      <th colspan=4>#22723 04a57e</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.71%</td>
      <td>84.71%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>82894</td>
      <td>82894</td>
      <td>-</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14958</td>
      <td>14958</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22723/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22723/coverage/framework)


## Modules coverage

### Stochastic tools

<table>
  <thead>
    <tr>
      <th></th>
      <th>b3bfe3</th>
      <th colspan=4>#22723 04a57e</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>91.11%</td>
      <td>91.11%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>6231</td>
      <td>6232</td>
      <td>+1</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>608</td>
      <td>608</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22723/coverage_diff/stochastic_tools)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22723/coverage/stochastic_tools)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/22723/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/22723/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/22723/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/22723/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/22723/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/22723/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/22723/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/22723/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/22723/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/22723/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/22723/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/22723/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/22723/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/22723/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/22723/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/22723/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/22723/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/22723/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/22723/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/22723/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/22723/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/22723/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/22723/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/22723/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/22723/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/22723/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/22723/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/22723/coverage/xfem)

</details>



This comment will be updated on new commits.

lindsayad commented on framework/src/utils/PetscSupport.C:
```diff
@@ -676,7 +682,7 @@ storePetscOptions(FEProblemBase & fe_problem, const InputParameters & params)
 
       if (option.first == "-matptap_via" || option.first == "-matmaijptap_via" ||
-          option.first == "-matproduct_ptap_via")
+          option.first == "-matproduct_ptap_via" || option.first == "-matproduct_ptap_via")
```
Are my eyes deceiving me? These look like the same string?

cticenhour replied:
They definitely are - should be `-mat_product_algorithm` instead. Thanks for the catch!
